### ISSUE TITLE
Simple oracle test doesn't seem to give good results 

### ISSUE THREAD

**Issue by kcrum** [swager/grf#247]
I'm generating simple linear data (in logit space), but the predicted TEs from `causal_forest` seem to be uncorrelated with the true oracle TEs. I would expect much better results than I'm currently getting. Here's my data-generating process:

```R
simple_oracle_data <- function(nsamp, beta=c(2,-1,0,0,0), te=1, hete=1, 
                               seed=3) {
  set.seed(seed)
  x <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(beta)))
  s <- rbinom(n=nsamp, size=1, prob=0.5)
  t <- rbinom(n=nsamp, size=1, prob=0.5)
  # x drives main effects, s drives HETE
  z <- x %*% beta + t*te + s*t*hete 
  # each example gets Z(t=0)
  z_ctl <- x %*% beta
  # each example gets Z(t=1)
  z_tmt <- x %*% beta + s*hete + te
  
  # Create tibble
  df <- tibble::as_tibble(x)
  # Rename "V#" to "x#" for numbered covariates
  names(df) <- gsub("V", "x", names(df))
  # Add subgroup indicator to df
  df[, "s"] <- as.double(s)
  # Add treatment indicator to df
  df[, "t"] <- as.factor(t)
    
  df[, "y"] <- as.double(rbinom(n=nsamp, size=1, prob=plogis(z)))
  df[, "p_true"] <- as.vector(plogis(z))
  df[, "p_ctl"] <- as.vector(plogis(z_ctl))
  df[, "p_tmt"] <- as.vector(plogis(z_tmt))
  df[, "te_true"] <- as.vector(plogis(z_tmt) - plogis(z_ctl))
  list("data" = df)
}
```

I generate train and test samples using this function, and then make predictions on the test set as follows:
```R
oracle_test <- function(ntrain=1000, ntest=10000, add_hete=TRUE, seed=7) {
  odtrain <- simple_oracle_data(nsamp = ntrain, seed = seed)
  # Pull out X, W, and Y
  xtrain <- model.matrix(~ ., data = select(odtrain$data, starts_with("x"), 
                                            starts_with("s")) )[, -1]
  wtrain <- as.numeric(odtrain$data$t) 
  ytrain <- as.numeric(odtrain$data$y) 
  # Train the model
  cf <-causal_forest(xtrain, wtrain, ytrain, seed=seed)
  
  # Get test data
  odtest <- simple_oracle_data(nsamp = ntest, seed = seed + 1)
  # Pull out X, W, and Y
  xtest <- model.matrix(~ ., data=select(odtest$data, starts_with("x"), 
                                         starts_with("s")))[, -1]
  ytest <- as.numeric(odtest$data$y) 
  ypred = predict(cf, xtest)
  cbind(tibble(ytrue = odtest$data$y, tetrue = odtest$data$te_true, 
               ypred = as.numeric(ypred$predictions)), 
        select(odtest$data, starts_with("s")))
}
```

When comparing the predicted TEs  in `ypred` to the known oracle TEs in the test set `tetrue`, I get a weak correlation:
```R
> cor(ot$tetrue, ot$ypred)
[1] 0.0984858
```
With a different random seed, this correlation comes out negative. Here's a scatter plot of the predictions from `causal_forest` versus the true TEs:

![image](https://user-images.githubusercontent.com/5655023/40400707-fa0644f6-5e08-11e8-9a14-37f54279f858.png)

I would expect it to be fairly easy for the forest to pick up heterogeneity in this case. I've checked things over fairly carefully, and a linear model returns reasonable coefficients:
```R
> summary(glm(y ~ x1 + x2 + x3 + x4 + x5 + s*t, data = odtrain$data, family = 'binomial'))

Call:
glm(formula = y ~ x1 + x2 + x3 + x4 + x5 + s * t, family = "binomial", 
    data = odtrain$data)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.4840  -0.6913   0.2513   0.6656   2.7131  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept) -0.07221    0.15444  -0.468    0.640    
x1           1.80200    0.12282  14.672  < 2e-16 ***
x2          -0.93426    0.09889  -9.448  < 2e-16 ***
x3           0.06619    0.08389   0.789    0.430    
x4           0.06071    0.08354   0.727    0.467    
x5          -0.03834    0.08456  -0.453    0.650    
s            0.34898    0.23363   1.494    0.135    
t1           1.20393    0.24501   4.914 8.93e-07 ***
s:t1         0.15643    0.34158   0.458    0.647    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1352.99  on 999  degrees of freedom
Residual deviance:  868.98  on 991  degrees of freedom
AIC: 886.98

Number of Fisher Scoring iterations: 5
```

It seems like the causal forest does a poor job of predicting treatment effect, however I'm fairly new to R *and* causal forestes, so even more than your typical issue, this is likely due to user error.

**Comment by kcrum** [comment]
For reference, when using the linear model from `glm` to produce a treatment effect estimate (i.e. predicting on all test points with t=1 and t=0), I get a correlation of 0.96.

**Comment by halflearned** [comment]
I will investigate this and get back to you all asap.

**Comment by kcrum** [comment]
Thank you @halflearned!

**Comment by halflearned** [comment]
@kcrum, thank you for your issue and for the minimal reproducible example. 

One problem seems to be that the function call was incorrect.

The code above reads
```
cf <-causal_forest(xtrain, wtrain, ytrain, seed=seed)
```

However, the correct argument order is 
```
cf <-causal_forest(xtrain, ytrain, wtrain, seed=seed)
```

Once we fix the order or (better) name the arguments, the correlation rises to around 0.67. With a training size of 10000, the correlation goes to 0.9. Moreover, if we (for diagnostics) shut off the nonlinearity in the true model, the correlation increases to about 0.96 ([gist](https://gist.github.com/halflearned/b727be706395d8c9dec027b946cf4bc1))

@swager and @susanathey, what do you think? A correlation of 0.67 is still a lot worse than glm. Does this warrant further investigation, or is it within the realm of acceptable?

**Comment by kcrum** [comment]
Gahhh, that's embarrassing. Thank you for looking into this!

I'm not surprised it's worse than glm, since glm is exactly the right model for the data generating process here.

**Comment by swager** [comment]
Thanks @halflearned, and no worries, @kcrum! On version 0.10.0, you may also get improved results using the `tune.parameters = TRUE` argument, which uses cross-validation for hyper-parameter search.
